book/src/dev/rfcs/xxxx-block-subsidy.md

```diff
@@ -0,0 +1,313 @@
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Given the module structure proposed above all the block subsidy calculations from the protocol must be implemented. The final goal is to do checks for each incoming block and make sure they pass the consensus rules.
+
+To do the calculations and checks the following constants, types and functions need to be introduced:
+
+## Constants
+
+- `SLOW_START_INTERVAL: Height`
+- `SLOW_START_SHIFT: Height`
+- `MAX_BLOCK_SUBSIDY: u64`
+- `PRE_BLOSSOM_POW_TARGET_SPACING: Duration`
+- `POST_BLOSSOM_POW_TARGET_SPACING: Duration`
+- `BLOSSOM_POW_TARGET_SPACING_RATIO: u64`
+- `PRE_BLOSSOM_HALVING_INTERVAL: Height`
+- `POST_BLOSSOM_HALVING_INTERVAL: Height`
+- `FOUNDERS_FRACTION_DIVISOR: u64`
+- `FOUNDERS_ADDRESS_COUNT: u32`
+- `FOUNDER_ADDRESSES_MAINNET: [&str; 48]`
+- `FOUNDER_ADDRESSES_TESTNET: [&str; 48]`
+
+### Funding streams parameter constants
+
+The design suggests to implement the parameters needed for funding streams as:
+
+```
+/// The funding stream receiver categories
+#[allow(missing_docs)]
+pub enum FundingStreamReceiver {
+    BootstrapProject,
+    ZcashFoundation,
+    MajorGrants,
+}
+
+/// The numerator for each funding stream receiving category 
+/// as described in [protocol specification §7.9.1][7.9.1].
+/// 
+/// [7.9.1]: https://zips.z.cash/protocol/protocol.pdf#zip214fundingstreams
+const FUNDING_STREAM_RECEIVER_NUMERATORS: &[(FundingStreamReceiver, u64)] = &[
+    (FundingStreamReceiver::BootstrapProject, 7),
+    (FundingStreamReceiver::ZcashFoundation, 5),
+    (FundingStreamReceiver::MajorGrants, 8),
+];
+
+/// Denominator as described in [protocol specification §7.9.1][7.9.1].
+/// 
+/// [7.9.1]: https://zips.z.cash/protocol/protocol.pdf#zip214fundingstreams
+pub const FUNDING_STREAM_RECEIVER_DENOMINATOR: u64 = 100;
+
+#[allow(missing_docs)]
+pub enum FundingStreamRange {
+    StartHeight,
+    EndHeight,
+}
+
+/// Start and end Heights for funding streams 
+/// as described in [protocol specification §7.9.1][7.9.1].
+/// 
+/// [7.9.1]: https://zips.z.cash/protocol/protocol.pdf#zip214fundingstreams
+const FUNDING_STREAM_HEIGHT_RANGES: &[(Network, Range<_>)] = &[
+    (Network::Mainnet, Height(1_046_400)..Height(2_726_400)),
+    (Network::Testnet, Height(1_028_500)..Height(2_796_000)),
+];
+```
+
+Each `FundingStreamReceiver` for the mainnet must have 48 payment addresses associated with and 51 for the testnet. 
+
+https://zips.z.cash/zip-0214#mainnet-recipient-addresses
+
+The following constants are needed:
+
+- `FUNDING_STREAM_BP_ADDRESSES_MAINNET: [&str; 48]`
+- `FUNDING_STREAM_ZF_ADDRESSES_MAINNET: [&str; 48]`
+- `FUNDING_STREAM_MG_ADDRESSES_MAINNET: [&str; 48]`
+- `FUNDING_STREAM_BP_ADDRESSES_TESTNET: [&str; 51]`
+- `FUNDING_STREAM_ZF_ADDRESSES_TESTNET: [&str; 51]`
+- `FUNDING_STREAM_MG_ADDRESSES_TESTNET: [&str; 51]`
```

Comment:
Some funding streams (currently ZF and MG) just use a single address for all 48 or 51 entries.

Let's implement them using `lazy_static!`, `Vec::repeat`, and `Vec::as_slice`?

(We don't need to update the RFC, because this is an implementation detail.)

Reply:
Actually I think we should update the RFC, because otherwise we'll forget to do it.

Reply:
It can be implemented as:

```
lazy_static! {
    // The Mainnet receivers for the BP slice.
    // As the 48 address are different they are enumerated one by one.
    pub static ref FUNDING_STREAM_BP_ADDRESSES_MAINNET: [Address; 48] = [
        Address::from_str("t3LmX1cxWPPPqL4TZHx42HU3U5ghbFjRiif").expect("address is valid"),
        Address::from_str("t3Toxk1vJQ6UjWQ42tUJz2rV2feUWkpbTDs").expect("address is valid"),
        ...
    ];

    // The Mainnet receivers for the ZF slice.
    // As the 48 address are the same the array is constructed by repeating the single address.
    pub static ref FUNDING_STREAM_ZF_ADDRESSES_MAINNET: [Address; 48] = {
        let single_address = "t3dvVE3SQEi7kqNzwrfNePxZ1d4hUyztBA1";
        std::iter::repeat(Address::from_str(single_address).expect("address is valid")).take(48).collect::<Vec<_>>()
    }.as_slice().try_into().expect("an array of addresses");

    // The Mainnet receivers for the MG slice.
    // As the 48 address are the same the array is constructed by repeating the single address.
    // Note: This can change in the future with the direct grant option considered in ZIP-214.
    pub static ref FUNDING_STREAM_MG_ADDRESSES_MAINNET: [Address; 48] = {
        let single_address = "t3XyYW8yBFRuMnfvm5KLGFbEVz25kckZXym";
        std::iter::repeat(Address::from_str(single_address).expect("address is valid")).take(48).collect::<Vec<_>>()
    }.as_slice().try_into().expect("an array of addresses");
}
```

This will probably work however there is probably a way to simplify it(too many `expect`s to me).

Let me know how do you see it.

Reply:
The `BP` string array will be easier to read and modify without the repeated conversion code. Once you have the array, you can convert it into addresses using:
```rust
    .iter()
    .map(|s| Address::from_str(s).expect(format!("address '{:?}' is valid", s)))
```

That way, you only write the conversion code once, and the invalid address is included in the `expect` panic message.

The single address cases look good - but please show the address in the `expect` panic message.